backend/mailru/mailru.go

```diff
@@ -0,0 +1,2356 @@
+package mailru
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	gohash "hash"
+	"io"
+	"path"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/rclone/rclone/backend/mailru/api"
+	"github.com/rclone/rclone/backend/mailru/mrhash"
+
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/config/configmap"
+	"github.com/rclone/rclone/fs/config/configstruct"
+	"github.com/rclone/rclone/fs/config/obscure"
+	"github.com/rclone/rclone/fs/fserrors"
+	"github.com/rclone/rclone/fs/fshttp"
+	"github.com/rclone/rclone/fs/hash"
+	"github.com/rclone/rclone/fs/object"
+	"github.com/rclone/rclone/fs/operations"
+
+	"github.com/rclone/rclone/lib/oauthutil"
+	"github.com/rclone/rclone/lib/pacer"
+	"github.com/rclone/rclone/lib/rest"
+
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+)
+
+// Global constants
+const (
+	minSleepPacer   = 10 * time.Millisecond
+	maxSleepPacer   = 2 * time.Second
+	decayConstPacer = 2          // bigger for slower decay, exponential
+	metaExpirySec   = 20 * 60    // meta server expiration time
+	serverExpirySec = 3 * 60     // download server expiration time
+	shardExpirySec  = 30 * 60    // upload server expiration time
+	maxServerLocks  = 4          // maximum number of locks per single download server
+	maxInt32        = 2147483647 // used as limit in directory list request
+	speedupMinSize  = 512        // speedup is not optimal if data is smaller than average packet
+)
+
+// Global errors
+var (
+	ErrorDirAlreadyExists   = errors.New("directory already exists")
+	ErrorDirSourceNotExists = errors.New("directory source does not exist")
+	ErrorInvalidName        = errors.New("invalid characters in object name")
+)
+
+// Description of how to authorize
+var oauthConfig = &oauth2.Config{
+	ClientID:     api.OAuthClientID,
+	ClientSecret: "",
+	Endpoint: oauth2.Endpoint{
+		AuthURL:   api.OAuthURL,
+		TokenURL:  api.OAuthURL,
+		AuthStyle: oauth2.AuthStyleInParams,
+	},
+}
+
+// Register with Fs
+func init() {
+	fs.Register(&fs.RegInfo{
+		Name:        "mailru",
+		Description: "Mail.ru Cloud",
+		NewFs:       NewFs,
+		Options: []fs.Option{{
+			Name:     "user",
+			Help:     "User name (usually email)",
+			Required: true,
+		}, {
+			Name:       "pass",
+			Help:       "Password",
+			Required:   true,
+			IsPassword: true,
+		}, {
+			Name:     "speedup_enable",
+			Default:  true,
+			Advanced: false,
+			Help: `Skip full upload if there is another file with same data hash.
+
+This feature is called "speedup" or "put by hash". It is especially efficient
+in case of generally available files like popular books, video or audio clips,
+because files are searched by hash in all accounts of all mailru users.
+Please note that rclone needs local memory and disk space to calculate
```

Comment:
Getting the mrhash into the hash framework will mean that you can ask the source for the mrhash which will mean you no longer need to buffer things to calculate the hash in advance.